kubeadm-token: search for existing kubeconfig files

neolit123 opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

Add means to search the current user home path and
the environment variable KUBECONFIG for existing files if the
user does not provide a --kubeconfig flag.

If the user provides a --kubeconfig flag respect it.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes/kubeadm/issues/198

**Special notes for your reviewer**:

i'm not 100% sure if that's the correct approach and if that's what's requested. so let's discuss it.

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Search standard KubeConfig file locations when using `kubeadm token` without `--kubeconfig`.
```

neolit123 committed:
kubeadm-token: search for existing kubeconfig files

Use clientcmd.NewDefaultClientConfigLoadingRules()
to search the standard locations for existing KubeConfig
files if the user did not provide a --kubeconfig flag.

If the user provides a --kubeconfig flag respect it.

Also do the following:
- Add a unit test in token_test.go for the
NewDefaultClientConfigLoadingRules() functionality.
- Rename the TestConfig and related constants to make
them unique to this file.

Fixes kubernetes/kubeadm#198

neolit123 commented:
@kad @liztio

neolit123 commented:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews

timothysc commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -381,3 +392,32 @@ func getClientset(file string, dryRun bool) (clientset.Interface, error) {
 	return kubeconfigutil.ClientSetFromFile(file)
 }
+
+func findExistingKubeConfig(file string) string {
```
Most of this code exists in clientcmd.NewNonInteractiveDeferredLoadingClientConfig and should not be leveraged using the client tooling.

xiangpengzhao commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -85,7 +92,7 @@ func NewCmdToken(out io.Writer, errW io.Writer) *cobra.Command {
 
 	tokenCmd.PersistentFlags().StringVar(&kubeConfigFile,
-		"kubeconfig", "/etc/kubernetes/admin.conf", "The KubeConfig file to use when talking to the cluster")
+		"kubeconfig", defaultKubeConfig, "The KubeConfig file to use when talking to the cluster. If the flag is not set the environment variable "+kubeConfigEnv+" and the user home directory and will be searched. If no existing KubeConfig files are found there the default location "+defaultKubeConfig+" will be used.")
```
>If the flag is not set the environment variable "+kubeConfigEnv+" and the user home directory and will be searched.

I don't understand this. Guess the last `and` is extra?

>If no existing KubeConfig files are found there the default location "+defaultKubeConfig+" will be used."

I'm not sure if `defaultKubeConfig` can be printed in this format. Let's see if the CI will be happy.

/ok-to-test

neolit123 replied:
yes, there is an extra `and` - typo.
i will probably replace this comment based on what `clientcmd.NewNonInteractiveDeferredLoadingClientConfig` (what a name...) does exactly behind the lines. on a quick look the `cmdutil.GetJoinCommand` call might have a problem with that...

timothysc reviewed (changes requested):
/assign @timothysc 
/approve 

Could you please add a UT then lgtm

neolit123 commented:
> Could you please add a UT then lgtm

ok, i need the other UT PR for `token` merged as a basis.

neolit123 commented:
link for my other PR that is needed before this:
https://github.com/kubernetes/kubernetes/pull/59220

neolit123 commented:
@timothysc 

update:
- added a unit test for the `NewDefaultClientConfigLoadingRules()` call using `kubeadm token delete` by setting a `KUBECONFIG` env. variable and not passing `--kubeconfig`.
- renamed some variables in `token_test.go` to make them more unique.

timothysc reviewed (approved):
/lgtm 

thx for the patch + update.

neolit123 commented:
/retest

neolit123 commented:
/meow

dims commented:
/retest

BenTheElder commented:
/ok-to-test

neolit123 commented:
/test pull-kubernetes-e2e-gce

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -50,6 +51,8 @@ import (
 	api "k8s.io/kubernetes/pkg/apis/core"
 )
 
+const defaultKubeConfig = "/etc/kubernetes/admin.conf"
```
Could you put this in `constants.go`?

luxas commented on cmd/kubeadm/app/cmd/token.go:
```diff
@@ -381,3 +387,16 @@ func getClientset(file string, dryRun bool) (clientset.Interface, error) {
+func findExistingKubeConfig(file string) string {
+	// The user did provide a --kubeconfig flag. Respect that and threat it as an
+	// explicit path without building a DefaultClientConfigLoadingRules object.
+	if file != defaultKubeConfig {
```
@neolit123 can you move this function to `cmd/kubeadm/app/cmd/util` please?
We have other places in kubeadm where it's used as a Kubernetes client tool, and we could/should leverage the same functionality to be consistent
